### PR TITLE
[error] add unittest for Error class

### PR DESCRIPTION
**Pull request by wgtdkp**
This PR adds unittest for the `Error` class.

### CODE REVIEW

File: src/common/error_test.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+}
+
+TEST_CASE("error-to-string", "[error]")
+{
+    SECTION("default Error object should return 'NONE' as string representation")
+    {
+        Error error;
+
+        REQUIRE(error.ToString() == "NONE");
```
Comment (simonlingoogle) [inline src/common/error_test.cpp:59]: - Use `OK` just like openthread?

Comment (wgtdkp) [reply src/common/error_test.cpp:59]: > Test string formatting for unsuccessful erros

Only `NONE` is documented for the `Error::ToString` API and string representation of other error are not documented and may change in the future. That means the user can reply on the truth that `Error(kNone).ToString()` always returns `NONE`, but no guarantee is provided for the rest errors.

> Test that error can be compared with error code
This has been tested by `TEST_CASE("error-comparison", "[error]")` at below.

> Test the behavior of comparing unsuccessful errors
I am not sure if we need to do, I am assuming that if comparison of one kind of error works, then the rest will work. Do we need to test if `ErrorA1 != ErrorB1`? There will be `17 * 17` testcases for this, and I think it is not necessary.

**Comment by wgtdkp** [comment]
> Test string formatting for unsuccessful erros

Only NONE is documented for the Error::ToString API and string representation of other error are not documented and may change in the future. That means the user can reply on the truth that Error(kNone).ToString() always returns NONE, but no guarantee is provided for the rest errors.

> Test that error can be compared with error code

This has been tested by TEST_CASE("error-comparison", "[error]") at below.

> Test the behavior of comparing unsuccessful errors

I am not sure if we need to do, I am assuming that if comparison of one kind of error works, then the rest will work. Do we need to test if ErrorA1 != ErrorB1? There will be 17 * 17 testcases for this, and I think it is not necessary.

File: src/common/error_test.cpp (hunk 1 above)
Comment (wgtdkp) [reply src/common/error_test.cpp:59]: ok ot me.

**Comment by simonlingoogle** [comment]
> Only NONE is documented for the Error::ToString API and string representation of other error are not documented and may change in the future. That means the user can reply on the truth that Error(kNone).ToString() always returns NONE, but no guarantee is provided for the rest errors.

Can we check that the `Error::ToString` returns `ErrorCodeToString(mCode) + ": " + mMessage)` ?
 
> > Test that error can be compared with error code
> 
> This has been tested by `TEST_CASE("error-comparison", "[error]")`.
> 
> > Test the behavior of comparing unsuccessful errors
> 
> I am not sure if we need to do, I am assuming that if comparison of one kind of error works, then the rest will work. Do we need to test if ErrorA1 != ErrorB1? There will be 17 * 17 testcases for this, and I think it is not necessary.

The tests only test for `OK`, but it may easily be the case that `OK` are implemented differently from `non-OK` errors. I think we should at lease test for one `non-OK` error for code coverage.

**Comment by wgtdkp** [comment]
> Can we check that the Error::ToString returns ErrorCodeToString(mCode) + ": " + mMessage) ?

I don't see a reason to test this, because we don't provide the format as a stable interface.

> The tests only test for OK, but it may easily be the case that OK are implemented differently from non-OK errors. I think we should at least test for one non-OK error for code coverage.

Since current implementation doesn't treat `OK` and `non-OK` errors, the error comparators are actually already covered by the unittest here. For future change that may implement `OK` error differently, it is the changer to write the additional unittest for it.

**Pull request merged**